src/routes/ui.rs

```diff
@@ -1,15 +1,168 @@
+use itertools::Itertools;
 use std::path::{Path, PathBuf};
+use std::borrow::Cow;
 use rocket::Route;
-use rocket::response::NamedFile;
+use rocket::http::{Cookie, Cookies};
+use rocket::request::{FlashMessage, Form};
+use rocket::response::{Flash, NamedFile, Redirect};
 use maud::{html, PreEscaped};
+use validator::Validate;
 
 use db;
-use db::models::Account;
+use db::models::{Account, NewAccount, NewUser, User};
 use templates::Page;
+use failure::Error;
 use error::Perhaps;
 
 pub fn routes() -> Vec<Route> {
-    routes![index, user_page, static_files]
+    routes![
+        index,
+        user_page,
+        auth_signin_get,
+        auth_signin_post,
+        auth_signout,
+        auth_signup_get,
+        auth_signup_post,
+        static_files
+    ]
+}
+
+#[get("/auth/sign_in")]
+pub fn auth_signin_get(flash: Option<FlashMessage>) -> Page {
+    Page::new().title("sign in").flash(flash).content(html! {
+        header h2 "sign in"
+
+        form method="post" {
+            div {
+                label for="username" "username:"
+                input type="text" id="username" name="username";
+            }
+
+            div {
+                label for="password" "password:"
+                input type="password" id="password" name="password";
+            }
+
+            button type="submit" "sign in"
+        }
+    })
+}
+
+#[derive(Debug, FromForm)]
+pub struct SigninForm {
+    username: String,
+    password: String,
+}
+
+#[post("/auth/sign_in", data = "<form>")]
+pub fn auth_signin_post(
+    form: Form<SigninForm>,
+    mut cookies: Cookies,
+    db_conn: db::Connection,
+) -> Result<Flash<Redirect>, Error> {
+    let form_data = form.get();
+    let user = User::by_username(&db_conn, &form_data.username)?;
+
+    if let Some(user) = user {
+        if user.valid_password(&form_data.password) {
+            cookies.add_private(Cookie::new("uid", user.id.to_string()));
```

Comment:
`add_private` is good, but I'm not sure where `secret_key` is set.  If we don't set one, Rocket will compute one at startup; however, this means that cookies won't be readable across server restarts.

Maybe it would be a good idea to add an example `Rocket.toml` with a `secret_key` that is rejected by default (so that one must be generated before starting Rustodon)?  This could probably be served nicely by its own issue, I guess.

Reply:
Oh, yes, that's a very good idea. We should also set a secret in production -- perhaps a `ROCKET_ environment variable override.

/cc @RobotDisco

Reply:
See #27.

Reply:
If PRODUCTION currently means what Travis deploys, we deffo have a place for this!